File: Fs.Units.CSharp/Simple/Mass.cs
// ReSharper disable once CheckNamespace
namespace Fs.Units.CSharp.Mass;

public static class MilligramExtensions
{
    public static double MilligramToGrams(this double x) => Units.Mass.Milligram.toGrams(x);
    public static double MilligramToKilograms(this double x) => Units.Mass.Milligram.toKilograms(x);
    public static double MilligramToPoundMass(this double x) => Units.Mass.Milligram.toPoundMass(x);
}

public static class GramExtensions
{
    public static double GramToMilligrams(this double x) => Units.Mass.Gram.toMilligrams(x);
    public static double GramToKilograms(this double x) => Units.Mass.Gram.toKilograms(x);
    public static double GramToPoundMass(this double x) => Units.Mass.Gram.toPoundMass(x);
}

public static class KilogramExtensions
{
    public static double KilogramToMilligrams(this double x) => Units.Mass.Kilogram.toMilligrams(x);
    public static double KilogramToGrams(this double x) => Units.Mass.Kilogram.toGrams(x);
    public static double KilogramToPoundMass(this double x) => Units.Mass.Kilogram.toPoundMass(x);
}

public static class PoundMassExtensions
{
    public static double PoundMassToMilligrams(this double x) => Units.Mass.PoundMass.toMilligrams(x);
    public static double PoundMassToGrams(this double x) => Units.Mass.PoundMass.toGrams(x);
    public static double PoundMassToKilograms(this double x) => Units.Mass.PoundMass.toKilograms(x);
}

File: src/Fs.Units/Simple/Mass.fs
namespace Fs.Units

[<AutoOpen>]
module Mass =

  /// Milligram
  [<Measure>]
  type mg

  /// Gram
  [<Measure>]
  type g

  /// Kilogram
  [<Measure>]
  type kg

  /// Pound mass
  [<Measure>]
  type lbm

  [<RequireQualifiedAccess>]
  module Milligram =

    let create (x: float) = x * 1.0<mg>
    let toKilograms (x: float<mg>) = x * 1e-6<kg> / 1.0<mg>
    let toGrams (x: float<mg>) = x * 1e-3<g> / 1.0<mg>
    let toPoundMass (x: float<mg>) = x * 1.0<lbm> / 453592.0<mg>

  [<RequireQualifiedAccess>]
  module Gram =

    let create (x: float) = x * 1.0<g>
    let toMilligrams (x: float<g>) = x * 1e3<mg> / 1.0<g>
    let toKilograms (x: float<g>) = x * 1e-3<kg> / 1.0<g>
    let toPoundMass (x: float<g>) = x * 1.0<lbm> / 453.592<g>

  [<RequireQualifiedAccess>]
  module Kilogram =

    let create (x: float) = x * 1.0<kg>
    let toMilligrams (x: float<kg>) = x * 1e6<mg> / 1.0<kg>
    let toGrams (x: float<kg>) = x * 1e3<g> / 1.0<kg>
    let toPoundMass (x: float<kg>) = x * 2.20462<lbm> / 1.0<kg>

  [<RequireQualifiedAccess>]
  module PoundMass =

    let create (x: float) = x * 1.0<lbm>
    let toMilligrams (x: float<lbm>) = x * 453592.0<mg> / 1.0<lbm>
    let toGrams (x: float<lbm>) = x * 453.592<g> / 1.0<lbm>
    let toKilograms (x: float<lbm>) = x * 1.0<kg> / 2.20462<lbm>
